lib/tree.rb

```diff
@@ -16,47 +16,127 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n) if balanced, otherwise O(n) where n is number of nodes
+  # Space Complexity: O(n), n is height
+  def add(key, value=nil)
```

Comment:
👍 Yeah recursion!  You threw me with the O(n) space complexity until I saw the note on height.